src/storage-client/src/controller/metrics_scraper.rs

```diff
@@ -0,0 +1,136 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A tokio task (and support machinery) for maintaining storage-managed
+//! collections.
+
+use futures::stream::BoxStream;
+use futures::StreamExt;
+
+use itertools::Itertools;
+use mz_orchestrator::ServiceProcessMetrics;
+use mz_ore::cast::CastFrom;
+use mz_repr::{Datum, GlobalId, Row};
+use timely::progress::ChangeBatch;
+use tokio::sync::oneshot;
+
+use crate::controller::collection_mgmt::CollectionManager;
+
+#[derive(Debug)]
+pub struct MetricsScraper {
+    shutdown_button: Option<oneshot::Sender<()>>,
+}
+
+impl Drop for MetricsScraper {
+    fn drop(&mut self) {
+        if let Some(shutdown_button) = self.shutdown_button.take() {
+            let _ = shutdown_button.send(());
+        }
+    }
+}
+
+/// Handle around a task that reads metrics from a stream and writes them to a
+/// storage collection via a handle to a `CollectionManager`.
+impl MetricsScraper {
+    /// Creates a new [`MetricsScraper`] that reads from the given
+    /// `metrics_stream` and writes it out to the collection identified by
+    /// `metrics_collection_id`.
+    ///
+    /// This assumes that the output collection is empty (or at least that
+    /// updates sum up to a zero diff) and that we are the sole writer. Once we
+    /// need support for concurrency we need to either teach `CollectionManager`
+    /// to deal with that or we need to learn to reconciliate our knowledge
+    /// about the contents of the output shard with other processes, by reading
+    /// back in the current contents of the output shard.
+    pub(super) fn new(
+        metrics_collection_id: GlobalId,
+        mut metrics_stream: BoxStream<
+            'static,
+            Vec<(
+                GlobalId,
+                String,
+                Result<Vec<ServiceProcessMetrics>, anyhow::Error>,
+            )>,
+        >,
+        collection_mgmt: CollectionManager,
+    ) -> Self {
+        let (shutdown_tx, mut shutdown_rx) = oneshot::channel();
+
+        mz_ore::task::spawn(|| "MetricsScraper", async move {
+            // Keep track of what we think the contents of the output
+            // collection, so that we can emit the required retractions/updates
+            // when we learn about new metrics.
+            //
+            // We assume that one update from the metrics stream contains
+            // entries for all known provisioned instances and if instances are
+            // not in an update their updates should be removed from the output
+            // collection.
+            let mut current_metrics = ChangeBatch::new();
+
+            loop {
+                tokio::select! {
+                    _msg = &mut shutdown_rx => {
+                        break;
+                    }
+
+                    Some(metrics) = metrics_stream.next() => {
```

Comment:
if its `None`, do you want to shutdown? or wait till the `shutdown_rx` to finish?

Reply:
Shutdown, I think. So I'll have to change this, right?